### PR TITLE
8329469: Generational ZGC: Move the methods forwarding_[index|find|insert] from zRelocate.cpp to ZForwarding

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch moves the methods `zRelocate.cpp/forwarding_[index|find|insert]`  from the file `zRelocate.cpp` to the class `ZForwarding`.  The first parameter of the methods `zRelocate.cpp/forwarding_[index|find|insert]` is always `ZForwarding*`. These methods should be placed in the class `ZForwarding` so that the first parameter `ZForwarding*` can be added by the compiler and then it can improve the encapsulation.

But the non-generational ZGC can't make such a move because the method signatures may be duplicate.

The test `make test-tier1_gc` passed locally. Thanks for taking the time to review.

Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329469](https://bugs.openjdk.org/browse/JDK-8329469): Generational ZGC: Move the methods forwarding_[index|find|insert] from zRelocate.cpp to ZForwarding (**Enhancement** - P5)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18579/head:pull/18579` \
`$ git checkout pull/18579`

Update a local copy of the PR: \
`$ git checkout pull/18579` \
`$ git pull https://git.openjdk.org/jdk.git pull/18579/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18579`

View PR using the GUI difftool: \
`$ git pr show -t 18579`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18579.diff">https://git.openjdk.org/jdk/pull/18579.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18579#issuecomment-2031588703)

### CODE REVIEW

File: src/hotspot/share/gc/z/zForwarding.hpp
Comment (stefank) [inline src/hotspot/share/gc/z/zForwarding.hpp:153]: Could some of these functions be moved to the `private:` section where you added the `index` function?

File: src/hotspot/share/gc/z/zRelocate.cpp
Comment (stefank) [inline src/hotspot/share/gc/z/zRelocate.cpp:356]: While looking at this, I now see that this could use the find overload that doesn't take a ZForwardingCursor. Could you update the code to do so?

File: src/hotspot/share/gc/z/zRelocate.cpp
Comment (stefank) [inline src/hotspot/share/gc/z/zRelocate.cpp:384]: Same here. The cursor is not necessary.

**Review by stefank — CHANGES_REQUESTED** [review]
This looks like a good cleanup. Some comments below:

File: src/hotspot/share/gc/z/zRelocate.cpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zRelocate.cpp]: Removed.

File: src/hotspot/share/gc/z/zForwarding.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zForwarding.hpp]: I moved four methods.

File: src/hotspot/share/gc/z/zRelocate.cpp
Comment (stefank) [reply src/hotspot/share/gc/z/zRelocate.cpp:356]: I agree. I missed that insert call in the diff.

**Comment by lgxbslgx** [comment]
What about this patch? Thanks.

File: src/hotspot/share/gc/z/zForwarding.hpp
Comment (stefank) [inline src/hotspot/share/gc/z/zForwarding.hpp:90]:
```suggestion
  uintptr_t index(zoffset from_offset);

  ZForwardingEntry find(uintptr_t from_index, ZForwardingCursor* cursor) const;
  zaddress find(zoffset from_offset, ZForwardingCursor* cursor);

  zoffset insert(uintptr_t from_index, zoffset to_offset, ZForwardingCursor* cursor);
  zaddress insert(zoffset from_offset, zaddress to_addr, ZForwardingCursor* cursor);
```

File: src/hotspot/share/gc/z/zForwarding.hpp
Comment (stefank) [inline src/hotspot/share/gc/z/zForwarding.hpp:152]:
```suggestion
  zaddress find(zaddress from_addr, ZForwardingCursor* cursor);
  zaddress find(zaddress_unsafe from_addr, ZForwardingCursor* cursor);
  zaddress find(zaddress_unsafe from_addr);

  zaddress insert(zaddress from_addr, zaddress to_addr, ZForwardingCursor* cursor);
```

**Review by stefank — CHANGES_REQUESTED** [review]
I think this looks good. I've added two whitespace suggestions below that I'd like to get incorporated. If you accept them, I'll mark this as Reviewed. Thanks.

File: src/hotspot/share/gc/z/zForwarding.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zForwarding.hpp]: Fixed.

File: src/hotspot/share/gc/z/zForwarding.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zForwarding.hpp]: Fixed.

**Review by stefank — APPROVED** [review]
Looks good.

**Comment by lgxbslgx** [comment]
@stefank Thanks for your review. Waiting for another review.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by lgxbslgx** [comment]
@stefank @fisk Thanks for your reviews. Integrating.

/integrate

**Pull request closed without merging**
